---
layout: vanilla
permalink: /papers/
description: "Papers describing Monocle."
modified: 2013-09-11
tags: [Monocle]
---
<!DOCTYPE html>
<html lang="{{ page.lang | default: site.lang | default: "en" }}">

{% include head.html %}

{% include header.html %}

<div class="container">
  <div class="row">
    <!-- sidebar, which will move to the top on a small screen -->
    <div class="col-sm-4">
      <nav id="toc" data-spy="affix" data-toggle="toc"></nav>
    </div>
    <!-- main content area -->
    <div class="col-sm-12">


    <div class="text-center">
      <h3 data-toc-skip>Cicero is an ongoing research project as well as a toolkit. <br>
        If you use Cicero, <strong>please cite these papers</strong> in your work! </h3>
    </div>

    <hr class="featurette-divider">

    <div class="row featurette">
      <div class="col-md-3">
        <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/cicero_diagram_fp.png" alt="Monocle 1">
      </div>

      <div class="col-md-9">
        <h3 data-toc-skip>Chromatin accessibility dynamics of myogenesis at single cell resolution</h3>
        <p>Hannah A. Pliner, Jonathan Packer, Jose McFaline-Figueroa, Darren Cusanovich, Riza Daza, Sanjay Srivatsan, Xiaojie Qiu, Dana Jackson, Anna Minkina, Andrew Adey, Frank Steemers, Jay Shendure*, Cole Trapnell*.
        <p><em>bioRxiv</em> 2017
        <br>
        <p> Over a million DNA regulatory elements have been cataloged in the human genome, but linking these elements to the genes that they regulate remains challenging. We introduce Cicero, a statistical method that connects regulatory elements to target genes using single cell chromatin accessibility data. We apply Cicero to investigate how thousands of dynamically accessible elements orchestrate gene regulation in differentiating myoblasts. Groups of co-accessible regulatory elements linked by Cicero meet criteria of “chromatin hubs”, in that they are physically proximal, interact with a common set of transcription factors, and undergo coordinated changes in histone marks that are predictive of gene expression. Pseudotemporal analysis revealed a subset of elements bound by MYOD in myoblasts that exhibit early opening, potentially serving as the initial sites of recruitment of chromatin remodeling and histone-modifying enzymes. The methodological framework described here constitutes a powerful new approach for elucidating the architecture, grammar and mechanisms of cis-regulation on a genome-wide basis.

        <div class="text-center">
          <a class="btn btn-default btn-sm" href="https://www.biorxiv.org/content/early/2017/06/26/155473" role="button">HTML</a>
          <a class="btn btn-default btn-sm" href="https://www.biorxiv.org/content/biorxiv/early/2017/06/26/155473.full.pdf" role="button">PDF</a>
        </div>

      </div>
    </div>
<br>

{% include footer.html %}


</body>

</html>
