---
layout: vanilla
permalink: /papers/
description: "Papers describing Monocle."
modified: 2013-09-11
tags: [Monocle]
---
<!DOCTYPE html>
<html lang="{{ page.lang | default: site.lang | default: "en" }}">

{% include head.html %}

{% include header.html %}

<div class="container">
  <div class="row">
    <!-- sidebar, which will move to the top on a small screen -->
    <div class="col-sm-4">
      <nav id="toc" data-spy="affix" data-toggle="toc"></nav>
    </div>
    <!-- main content area -->
    <div class="col-sm-12">


    <div class="text-center">
      <h3 data-toc-skip>Cicero is an ongoing research project as well as a toolkit. <br>
        If you use Cicero, <strong>please cite these papers</strong> in your work! </h3>
    </div>

    <hr class="featurette-divider">

    <div class="row featurette">
      <div class="col-md-3">
        <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/fx1_lrg_cicero.jpg" alt="Monocle 1">
      </div>

      <div class="col-md-9">
        <h3 data-toc-skip>Cicero Predicts <i>cis</i>-Regulatory DNA Interactions from Single-Cell Chromatin Accessibility Data</h3>
        <p>Hannah A. Pliner, Jonathan S. Packer, José L. McFaline-Figueroa, Darren A. Cusanovich, Riza M. Daza, Sanjay Srivatsan, Xiaojie Qiu, Dana Jackson, Anna Minkina, Andrew C. Adey, Frank J. Steemers, Jay Shendure*, Cole Trapnell*.
        <p><em>Molecular Cell</em> 71, 1–14 2018
        <br>
        <p> Linking regulatory DNA elements to their target genes, which may be located hundreds of kilobases away, remains challenging. Here, we introduce Cicero, an algorithm that identifies co-accessible pairs of DNA elements using single-cell chromatin accessibility data and so connects regulatory elements to their putative target genes. We apply Cicero to investigate how dynamically accessible elements orchestrate gene regulation in differentiating myoblasts. Groups of Cicero-linked regulatory elements meet criteria of “chromatin hubs”—they are enriched for physical proximity, interact with a common set of transcription factors, and undergo coordinated changes in histone marks that are predictive of changes in gene expression. Pseudotemporal analysis revealed that most DNA elements remain in chromatin hubs throughout differentiation. A subset of elements bound by MYOD1 in myoblasts exhibit early opening in a PBX1- and MEIS1-dependent manner. Our strategy can be applied to dissect the architecture, sequence determinants, and mechanisms of cis-regulation on a genome-wide scale.
        </p>
        <div class="text-center">
          <a class="btn btn-default btn-sm" href="https://www.cell.com/molecular-cell/fulltext/S1097-2765(18)30547-1#%20" role="button">HTML</a>
          <a class="btn btn-default btn-sm" href="https://www.cell.com/action/showPdf?pii=S1097-2765%2818%2930547-1" role="button">PDF</a>
        </div>
      </div>

      <div class="row featurette">
        <div class="col-md-3">
          <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/fx1_lrg.jpg" alt="Monocle 1">
        </div>
        <div class="col-md-9">
          <h3 data-toc-skip>A Single-Cell Atlas of In Vivo Mammalian Chromatin Accessibility</h3>
          <p>Darren A. Cusanovich, Andrew J. Hill, Delasa Aghamirzaie, Riza M. Daza, Hannah A. Pliner, Joel B. Berletch, Galina N. Filippova, Xingfan Huang, Lena Christiansen, William S. DeWitt, Choli Lee, Samuel G. Regalado, David F. Read, Frank J. Steemers, Christine M. Disteche, Cole Trapnell, Jay Shendure.
          <p><em>Cell</em> 174, 1–16 2018
          <br>
          <p> We applied a combinatorial indexing assay, sci-ATAC-seq, to profile genome-wide chromatin accessibility in ∼100,000 single cells from 13 adult mouse tissues. We identify 85 distinct patterns of chromatin accessibility, most of which can be assigned to cell types, and ∼400,000 differentially accessible elements. We use these data to link regulatory elements to their target genes, to define the transcription factor grammar specifying each cell type, and to discover in vivo correlates of heterogeneity in accessibility within cell types. We develop a technique for mapping single cell gene expression data to single-cell chromatin accessibility data, facilitating the comparison of atlases. By intersecting mouse chromatin accessibility with human genome-wide association summary statistics, we identify cell-type-specific enrichments of the heritability signal for hundreds of complex traits. These data define the in vivo landscape of the regulatory genome for common mammalian cell types at single-cell resolution.
          </p>
          <div class="text-center">
            <a class="btn btn-default btn-sm" href="https://www.cell.com/cell/abstract/S0092-8674(18)30855-9" role="button">HTML</a>
            <a class="btn btn-default btn-sm" href="https://www.cell.com/action/showPdf?pii=S0092-8674%2818%2930855-9" role="button">PDF</a>
          </div>
        </div>      
    </div>
<br>

{% include footer.html %}


</body>

</html>
