---
layout: vanilla
permalink: /papers/
description: "Papers describing Monocle."
modified: 2013-09-11
tags: [Monocle]
---
<!DOCTYPE html>
<html lang="{{ page.lang | default: site.lang | default: "en" }}">

{% include head.html %}

{% include header.html %}

<div class="container">
  <div class="row">
    <!-- sidebar, which will move to the top on a small screen -->
    <div class="col-sm-4">
      <nav id="toc" data-spy="affix" data-toggle="toc"></nav>
    </div>
    <!-- main content area -->
    <div class="col-sm-12">


    <div class="text-center">
      <h3 data-toc-skip>Cicero is an ongoing research project as well as a toolkit. <br>
        If you use Cicero, <strong>please cite these papers</strong> in your work! </h3>
    </div>

    <hr class="featurette-divider">

    <div class="row featurette">
      <div class="col-md-3">
        <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/cicero_diagram_fp.png" alt="Monocle 1">
      </div>

      <div class="col-md-9">
        <h3 data-toc-skip>Cicero Predicts <i>cis</i>-Regulatory DNA Interactions from Single-Cell Chromatin Accessibility Data</h3>
        <p>Hannah A. Pliner, Jonathan S. Packer, José L. McFaline-Figueroa, Darren A. Cusanovich, Riza M. Daza, Sanjay Srivatsan, Xiaojie Qiu, Dana Jackson, Anna Minkina, Andrew C. Adey, Frank J. Steemers, Jay Shendure*, Cole Trapnell*.
        <p><em>Molecular Cell</em> 71, 1–14 2018
        <br>
        <p> Linking regulatory DNA elements to their target genes, which may be located hundreds of kilobases away, remains challenging. Here, we introduce Cicero, an algorithm that identifies co-accessible pairs of DNA elements using single-cell chromatin accessibility data and so connects regulatory elements to their putative target genes. We apply Cicero to investigate how dynamically accessible elements orchestrate gene regulation in differentiating myoblasts. Groups of Cicero-linked regulatory elements meet criteria of “chromatin hubs”—they are enriched for physical proximity, interact with a common set of transcription factors, and undergo coordinated changes in histone marks that are predictive of changes in gene expression. Pseudotemporal analysis revealed that most DNA elements remain in chromatin hubs throughout differentiation. A subset of elements bound by MYOD1 in myoblasts exhibit early opening in a PBX1- and MEIS1-dependent manner. Our strategy can be applied to dissect the architecture, sequence determinants, and mechanisms of cis-regulation on a genome-wide scale.
        </p>
        <div class="text-center">
          <a class="btn btn-default btn-sm" href="https://www.cell.com/molecular-cell/fulltext/S1097-2765(18)30547-1#%20" role="button">HTML</a>
          <a class="btn btn-default btn-sm" href="https://www.cell.com/action/showPdf?pii=S1097-2765%2818%2930547-1" role="button">PDF</a>
        </div>
      </div>
    </div>
<br>

{% include footer.html %}


</body>

</html>
