---
title: "Single Cell ATAC-seq Analysis with Cicero"
author: "Hannah Pliner"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    theme: united
vignette: >
  %\VignetteIndexEntry{Vignette from Cicero Website}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

  This vignette is a condensed version of the documentation pages on the 
  [Cicero website](https://cole-trapnell-lab.github.io/cicero-release). Please 
  check out the website for more details.
  
  **WARNING:** This vignette demonstrates usage with the new 
  [Monocle 3](https://cole-trapnell-lab.github.io/monocle3) version of Cicero. 
  More information about the Cicero transition from Monocle 2 to Monocle 3 is 
  available on the 
  [Cicero website](https://cole-trapnell-lab.github.io/cicero-release).
  
# Introduction:
  The main purpose of Cicero is to use single-cell chromatin accessibility data 
  to predict regions of the genome that are more likely to be in physical 
  proximity in the nucleus. This can be used to identify putative 
  enhancer-promoter pairs, and to get a sense of the overall stucture of the 
  cis-architecture of a genomic region.

  Because of the sparsity of single-cell data, cells must be aggregated by 
  similarity to allow robust correction for various technical factors in the 
  data.

  Ultimately, Cicero provides a "Cicero co-accessibility" score between -1 and 
  1 between each pair of accessible peaks within a user defined distance where 
  a higher score indicates higher co-accessibility.

  In addition, the Cicero package provides an extension toolkit for analyzing 
  single-cell ATAC-seq experiments using the framework provided by the 
  single-cell RNA-seq analysis software, 
  [Monocle 3](https://cole-trapnell-lab.github.io/monocle3). This vignette 
  provides an overview of a single-cell ATAC-Seq analysis workflow with Cicero. 
  For further information and more options, please see the manual pages for the 
  Cicero R package, the 
  [Cicero website](https://cole-trapnell-lab.github.io/cicero-release) and our 
  publications.

  Cicero can help you perform two main types of analysis:

  * **Constructing and analyzing cis-regulatory networks.** Cicero analyzes 
  co-accessibility to identify putative cis-regulatory interactions, and uses 
  various techniques to visualize and analyze them.
  * **General single-cell chromatin accessibility analysis.** Cicero also 
  extends the software package Monocle 3 to allow for identification of 
  differential accessibility, clustering, visualization, and trajectory 
  reconstruction using single-cell chromatin accessibility data.
  
# Installing Cicero

1. Download and install Monocle 3 following the instructions [here](https://cole-trapnell-lab.github.io/monocle3/monocle3_docs/#installing-monocle-3) 
and the Monocle 3 branch of Cicero from Github:

```{r, eval = FALSE}
devtools::install_github(cole-trapnell-lab/cicero-release, ref = "monocle3")
```
 
2. Load the package into R session.
```{r Load, message=FALSE}
library(monocle3)
library(cicero)
```

# Constructing cis-regulatory networks

## Running Cicero

### The cell_data_set class
Cicero holds data in objects of the cell_data_set (CDS) class. The class is 
derived from the Bioconductor SingleCellExperiment class, which provides a 
common interface familiar to those who have analyzed single-cell experiments 
with Bioconductor. Monocle 3 provides detailed documentation about how to 
generate an input CDS 
[here](https://cole-trapnell-lab.github.io/monocle3/monocle3_docs/#the-cell-data-set-class).

To modify the CDS object to hold chromatin accessibility rather than expression 
data, Cicero uses peaks as its feature data (fData) rather than genes or 
transcripts. Specifically, many Cicero functions require peak information in the 
form chr1_10390134_10391134. For example, an input fData table might look like 
this:

&nbsp;                    |	site_name                 |	chromosome|	bp1	      | bp2
--------------------------|---------------------------|-----------|-----------|------------
chr10_100002625_100002940 |	chr10_100002625_100002940 |	10        | 100002625 | 100002940
chr10_100006458_100007593 |	chr10_100006458_100007593 |	10        | 100006458 | 100007593
chr10_100011280_100011780 |	chr10_100011280_100011780 |	10        | 100011280 | 100011780
chr10_100013372_100013596 |	chr10_100013372_100013596 |	10        | 100013372 | 100013596
chr10_100015079_100015428 |	chr10_100015079_100015428 |	10        | 100015079 | 100015428

The you can access an example dataset derived from the [Mouse sci-ATAC-seq Atlas](http://atlas.gs.washington.edu/mouse-atac/) published in [Cusanovich and Hill, et. al.](https://www.cell.com/cell/fulltext/S0092-8674(18)30855-9) using the code below:

```{r}
# Code to download (54M) and unzip the file - can take a couple minutes 
# depending on internet connection:
temp <- textConnection(readLines(gzcon(url("http://staff.washington.edu/hpliner/data/kidney_data.txt.gz"))))

# read in the data
cicero_data <- read.table(temp)
```

For convenience, Cicero includes a function called make_atac_cds. This function 
takes as input a data.frame or a path to a file in a sparse matrix format. 
Specifically, this file should be a tab-delimited text file with three columns. 
The first column is the peak coordinates in the form 
"chr10_100013372_100013596", the second column is the cell name, and the third 
column is an integer that represents the number of reads from that cell 
overlapping that peak. The file should not have a header line.  

For example:

&nbsp;                    | &nbsp; | &nbsp;
--------------------------|--------|---------
chr10_100002625_100002940 |	cell1  | 1
chr10_100006458_100007593 |	cell2  | 2
chr10_100006458_100007593 |	cell3  | 1
chr10_100013372_100013596 |	cell2  | 1
chr10_100015079_100015428 |	cell4  | 3

The output of make_atac_cds is a valid CDS object ready to be input into 
downstream Cicero functions.

```{r, eval=TRUE}
input_cds <- make_atac_cds(cicero_data, binarize = TRUE)
```

### Create a Cicero CDS

Because single-cell chromatin accessibility data is extremely sparse, accurate 
estimation of co-accessibility scores requires us to aggregate similar cells to 
create more dense count data. Cicero does this using a k-nearest-neighbors 
approach which creates overlapping sets of cells. Cicero constructs these sets 
based on a reduced dimension coordinate map of cell similarity, for example, 
from a tSNE or UMAP map.

You can use any dimensionality reduction method to base your aggregated CDS on. 
We will show you how to create two versions, tSNE and UMAP (below). Both of 
these dimensionality reduction methods are available from 
[Monocle 3](http://cole-trapnell-lab.github.io/monocle3/) (and loaded 
by Cicero).

Once you have your reduced dimension coordinate map, you can use the function 
make_cicero_cds to create your aggregated CDS object. The input to 
make_cicero_cds is your input CDS object, and your reduced dimension coordinate 
map. The reduced dimension map reduced_coordinates should be in the form of a 
data.frame or a matrix where the row names match the cell IDs from the pData 
table of your CDS. The columns of reduced_coordinates should be the coordinates 
of the reduced dimension object, for example:

&nbsp;   	    | umap_coord1    | umap_coord2
--------------|----------------|---------------
cell1         | -0.7084047     | -0.7232994
cell2         |	-4.4767964     | 0.8237284
cell3         |	1.4870098      | -0.4723493

Here is an example of both dimensionality reduction and creation of a Cicero 
CDS. Using Monocle 3 as a guide, we first find UMAP coordinates for our 
input_cds:

```{r, eval=TRUE}
set.seed(2017)
input_cds <- detect_genes(input_cds)
input_cds <- estimate_size_factors(input_cds)
input_cds <- preprocess_cds(input_cds, method = "LSI")
input_cds <- reduce_dimension(input_cds, reduction_method = 'UMAP', 
                              preprocess_method = "LSI")
```

For more information on the above code, see the 
[Monocle 3](http://cole-trapnell-lab.github.io/monocle3/) website 
section on clustering cells.

We can use Monocle's plotting function to visualize the reduced dimension map:
```{r}
plot_cells(input_cds)
```

Next, we access the UMAP coordinates from the input CDS object where they are 
stored by Monocle 3 and run make_cicero_cds:

```{r, eval=TRUE}
umap_coords <- reducedDims(input_cds)$UMAP
cicero_cds <- make_cicero_cds(input_cds, reduced_coordinates = umap_coords)
```

### Run Cicero

The main function of the Cicero package is to estimate the co-accessiblity of 
sites in the genome in order to predict cis-regulatory interactions. There are 
two ways to get this information:

* **run_cicero, get Cicero outputs with all defaults** The function run_cicero will 
call each of the relevant pieces of Cicero code using default values, and 
calculating best-estimate parameters as it goes. For most users, this will be 
the best place to start.
* **Call functions separately, for more flexibility** For users wanting more 
flexibility in the parameters that are called, and those that want access to 
intermediate information, Cicero allows you to call each of the component parts 
separately. More information about running function separately is available on 
the package manual pages and on the 
[Cicero website](https://cole-trapnell-lab.github.io/cicero-release).

The easiest way to get Cicero co-accessibility scores is to run run_cicero. To 
run run_cicero, you need a cicero CDS object (created above) and a genome 
coordinates file, which contains the lengths of each of the chromosomes in your 
organism. The human hg19 coordinates and the mouse mm9 coordinates are included 
with the package and can be accessed with data("human.hg19.genome") and 
data("mouse.mm9.genome"). Here is an example call, continuing with our example 
data:

**Note** for speed in the vignette, we run Cicero here on only a small portion 
of chromosome 2. 

```{r, eval=TRUE}
data("mouse.mm9.genome")

# use only a small part of the genome for speed
sample_genome <- subset(mouse.mm9.genome, V1 == "chr2")
sample_genome$V2[1] <- 10000000

conns <- run_cicero(cicero_cds, sample_genome, sample_num = 2) # Takes a few minutes to run
head(conns)
```

## Visualizing Cicero Connections
The Cicero package includes a general plotting function for visualizing 
co-accessibility called plot_connections. This function uses the Gviz framework 
for plotting genome browser-style plots. We have adapted a function from the 
Sushi R package for mapping connections. plot_connections has many options, 
some detailed in the Advanced Visualization section on the 
[Cicero website](https://cole-trapnell-lab.github.io/cicero-release), but 
to get a basic plot from your co-accessibility table is quite simple. 

Here, we will include optional gene_model data so that genes can be plotted as
well. 

```{r}
# Download the GTF associated with this data (mm9) from ensembl and load it
# using rtracklayer

# download and unzip
temp <- tempfile()
download.file("ftp://ftp.ensembl.org/pub/release-65/gtf/mus_musculus/Mus_musculus.NCBIM37.65.gtf.gz", temp)
gene_anno <- rtracklayer::readGFF(temp)
unlink(temp)

# rename some columns to match requirements
gene_anno$chromosome <- paste0("chr", gene_anno$seqid)
gene_anno$gene <- gene_anno$gene_id
gene_anno$transcript <- gene_anno$transcript_id
gene_anno$symbol <- gene_anno$gene_name

```

```{r, fig.width = 7, fig.height = 4, fig.align='center', eval=TRUE}

plot_connections(conns, "chr2", 9773451, 9848598,
                 gene_model = gene_anno, 
                 coaccess_cutoff = .25, 
                 connection_width = .5, 
                 collapseTranscripts = "longest" )
```

## Comparing Cicero connections to other datasets
Often, it is useful to compare Cicero connections to other datasets with 
similar kinds of links. For example, you might want to compare the output of 
Cicero to ChIA-PET ligations. To do this, Cicero includes a function called 
compare_connections. This function takes as input two data frames of connection 
pairs, conns1 and conns2, and returns a logical vector of which connections from 
conns1 are found in conns2. The comparison in this function is conducted using 
the GenomicRanges package, and uses the max_gap argument from that package to 
allow slop in the comparisons.

For example, if we wanted to compare our Cicero predictions to a set of 
(made-up) ChIA-PET connections, we could run:

```{r, eval=TRUE}
chia_conns <-  data.frame(Peak1 = c("chr2_3005100_3005200", "chr2_3004400_3004600", 
                                    "chr2_3004900_3005100"), 
                          Peak2 = c("chr2_3006400_3006600", "chr2_3006400_3006600", 
                                    "chr2_3035100_3035200"))
head(chia_conns)

conns$in_chia <- compare_connections(conns, chia_conns)

head(conns)
```

You may find that this overlap is too strict when comparing completely distinct 
datasets. Looking carefully, the 3rd line of the ChIA-PET data matches fairly 
closely to the last line shown of conns. The difference is only ~80 base pairs, 
which could be a matter of peak-calling. This is where the max_gap parameter 
can be useful:

```{r, eval=TRUE}
conns$in_chia_100 <- compare_connections(conns, chia_conns, maxgap=100)

head(conns)
```

In addition, Cicero's plotting function has a way to compare datasets visually. 
To do this, use the comparison_track argument. The comparison data frame must 
include a third columns beyond the first two peak columns called "coaccess". 
This is how the plotting function determines the height of the plotted 
connections. This could be a quantitative measure, like the number of ligations 
in ChIA-PET, or simply a column of 1s. More info on plotting options in manual 
pages ?plot_connections and in the Advanced Visualization section of the 
[Cicero website](https://cole-trapnell-lab.github.io/cicero-release).

```{r, fig.width = 7, fig.height = 4, fig.align='center', eval=TRUE}
# Add a column of 1s called "coaccess"
chia_conns <-  data.frame(Peak1 = c("chr2_3005100_3005200", "chr2_3004400_3004600", 
                                    "chr2_3004900_3005100"), 
                          Peak2 = c("chr2_3006400_3006600", "chr2_3006400_3006600", 
                                    "chr2_3035100_3035200"),
                          coaccess = c(1, 1, 1))

plot_connections(conns, "chr2", 3004000, 3040000, 
                 gene_model = gene_anno, 
                 coaccess_cutoff = 0,
                 connection_width = .5,
                 comparison_track = chia_conns,
                 include_axis_track = FALSE,
                 collapseTranscripts = "longest") 
```


## Finding Cis-Coaccessibility Networks (CCANS)
In addition to pairwise co-accessibility scores, Cicero also has a function to 
find Cis-Co-accessibility Networks (CCANs), which are modules of sites that are 
highly co-accessible with one another. We use the Louvain community detection 
algorithm (Blondel et al., 2008) to find clusters of sites that tend to be 
co-accessible. The function generate_ccans takes as input a connection data 
frame and outputs a data frame with CCAN assignments for each input peak. Sites 
not included in the output data frame were not assigned a CCAN.

The function generate_ccans has one optional input called 
coaccess_cutoff_override. When coaccess_cutoff_override is NULL, the function 
will determine and report an appropriate co-accessibility score cutoff value 
for CCAN generation based on the number of overall CCANs at varying cutoffs. 
You can also set coaccess_cutoff_override to be a numeric between 0 and 1, to 
override the cutoff-finding part of the function. This option is useful if you 
feel that the cutoff found automatically was too strict or loose, or for speed 
if you are rerunning the code and know what the cutoff will be, since the 
cutoff finding procedure can be slow.

```{r, eval=TRUE}
CCAN_assigns <- generate_ccans(conns)

head(CCAN_assigns)
```

## Cicero gene activity scores

We have found that often, accessibility at promoters is a poor predictor of 
gene expression. However, using Cicero links, we are able to get a better sense 
of the overall accessibility of a promoter and it's associated distal sites. 
This combined score of regional accessibility has a better concordance with 
gene expression. We call this score the Cicero gene activity score, and it is 
calculated using two functions.

The initial function is called build_gene_activity_matrix. This function takes 
an input CDS and a Cicero connection list, and outputs an unnormalized table of 
gene activity scores. **IMPORTANT**: the input CDS must have a column in the 
fData table called "gene" which indicates the gene if that peak is a promoter, 
and NA if the peak is distal. One way to add this column is demonstrated below.

The output of build_gene_activity_matrix is unnormalized. It must be normalized 
using a second function called normalize_gene_activities. If you intend to 
compare gene activities across different datasets of subsets of data, then all 
gene activity subsets should be normalized together, by passing in a list of 
unnormalized matrices. If you only wish to normalized one matrix, simply pass 
it to the function on its own. normalize_gene_activities also requires a named 
vector of of total accessible sites per cell. This is easily found in the pData 
table of your CDS, called "num_genes_expressed". See below for an example.

```{r, eval=TRUE}

#### Add a column for the pData table indicating the gene if a peak is a promoter ####
# Create a gene annotation set that only marks the transcription start sites of 
# the genes. We use this as a proxy for promoters.
# To do this we need the first exon of each transcript
pos <- subset(gene_anno, strand == "+")
pos <- pos[order(pos$start),] 
pos <- pos[!duplicated(pos$transcript),] # remove all but the first exons per transcript
pos$end <- pos$start + 1 # make a 1 base pair marker of the TSS

neg <- subset(gene_anno, strand == "-")
neg <- neg[order(neg$start, decreasing = TRUE),] 
neg <- neg[!duplicated(neg$transcript),] # remove all but the first exons per transcript
neg$start <- neg$end - 1

gene_annotation_sub <- rbind(pos, neg)

# Make a subset of the TSS annotation columns containing just the coordinates 
# and the gene name
gene_annotation_sub <- gene_annotation_sub[,c("chromosome", "start", "end", "symbol")]

# Rename the gene symbol column to "gene"
names(gene_annotation_sub)[4] <- "gene"

input_cds <- annotate_cds_by_site(input_cds, gene_annotation_sub)

tail(fData(input_cds))

#### Generate gene activity scores ####
# generate unnormalized gene activity matrix
unnorm_ga <- build_gene_activity_matrix(input_cds, conns)

# remove any rows/columns with all zeroes
unnorm_ga <- unnorm_ga[!Matrix::rowSums(unnorm_ga) == 0, !Matrix::colSums(unnorm_ga) == 0]

# make a list of num_genes_expressed
num_genes <- pData(input_cds)$num_genes_expressed
names(num_genes) <- row.names(pData(input_cds))

# normalize
cicero_gene_activities <- normalize_gene_activities(unnorm_ga, num_genes)

# if you had two datasets to normalize, you would pass both:
# num_genes should then include all cells from both sets
unnorm_ga2 <- unnorm_ga
cicero_gene_activities <- normalize_gene_activities(list(unnorm_ga, unnorm_ga2), num_genes)

```

# Single-cell accessibility trajectories
The second major function of the Cicero package is to extend Monocle 3 for use 
with single-cell accessibility data. The main obstacle to overcome with 
chromatin accessibility data is the sparsity, so most of the extensions and 
methods are designed to address that.

## Constructing trajectories with accessibility data
We strongly recommend that you consult the Monocle 3 website, especially 
[this section](https://cole-trapnell-lab.github.io/monocle3/monocle3_docs/#constructing-single-cell-trajectories)
prior to reading about Cicero's extension of the Monocle analysis described. 
Briefly, Monocle 3 infers pseudotime trajectories in five steps:

1. Preprocess the data
2. Reduce the dimensionality of the data
3. Cluster the cells
4. Learn the trajectory graph
5. Order the cells in pseudotime

These steps can be run on accessibility data with only minor modifications:

First, we download and load in the data (same as above):
```{r}
# Code to download (54M) and unzip the file - can take a couple minutes 
# depending on internet connection:
temp <- textConnection(readLines(gzcon(url("http://staff.washington.edu/hpliner/data/kidney_data.txt.gz"))))

# read in the data
cicero_data <- read.table(temp)
```

Next, we preprocess the data using Latent Semantic Indexing (LSI) and then 
continue with the standard dimensionality reduction used in Monocle 3. See
[here](http://andrewjohnhill.com/blog/2019/05/06/dimensionality-reduction-for-scatac-data/) 
for an excellent discussion by Andrew Hill on why we use LSI for single-cell 
ATAC-seq.
```{r}
set.seed(2017)
input_cds <- estimate_size_factors(input_cds)

#1
input_cds <- preprocess_cds(input_cds, method = "LSI")

#2
input_cds <- reduce_dimension(input_cds, reduction_method = 'UMAP', 
                              preprocess_method = "LSI")

#3
input_cds <- cluster_cells(input_cds)

#4
input_cds <- learn_graph(input_cds)

#5
# cell ordering can be done interactively by leaving out "root_cells"
input_cds <- order_cells(input_cds, root_cells = "GAGATTCCAGTTGAATCACTCCATCGAGATAGAGGC")
```

Plot the results:
```{r, fig.width = 7, fig.height = 4, fig.align='center', eval=TRUE}

plot_cells(input_cds, color_cells_by = "pseudotime")
```


### Aggregation: the primary method for addressing sparsity
The primary way that the Cicero package deals with the sparsity of single-cell 
chromatin accessibility data is through aggregation. Aggregating the counts of 
either single cells or single peaks allows us to produce a "consensus" count 
matrix, reducing noise and allowing us to move out of the binary regime. Under 
this grouping, the number of cells in which a particular site is accessible can 
be modeled with a binomial distribution or, for sufficiently large groups, the 
corresponding Gaussian approximation. Modeling grouped accessibility counts as 
normally distributed allows Cicero to easily adjust them for arbitrary technical 
covariates by simply fitting a linear model and taking the residuals with 
respect to it as the adjusted accessibility score for each group of cells. We 
demonstrate how to apply this grouping practically below.


## Differential Accessibility Analysis
Once you have your cells ordered in pseudotime, you can ask where in the genome 
chromatin accessibility is changing across pseudotime. If you know of specific sites 
that are important to your system, you may want to visualize the accessibility 
at those sites across pseudotime.

### Visualizing accessibility across pseudotime

For simplicity, we will include only the main section of the trajectory.

```{r, fig.width = 3, fig.height = 4, fig.align='center', eval=TRUE}
input_cds_lin <- input_cds[,is.finite(pseudotime(input_cds))]

plot_accessibility_in_pseudotime(input_cds_lin[c("chr1_3238849_3239700", 
                                                 "chr1_3406155_3407044", 
                                                 "chr1_3397204_3397842")])
```


### Running fit models with single-cell chromatin accessibility data
In this section, we are interested in a 
site-level statistic (whether a site is changing in pseudotime), so we will 
aggregate similar cells. To do this, Cicero has a useful function called 
aggregate_by_cell_bin.

#### aggregate_by_cell_bin
We use the function aggregate_by_cell_bin to aggregate our input CDS object by 
a column in the pData table. In this example, we will assign cells to bins by 
cutting the pseudotime trajectory into 10 parts.

```{r, eval=TRUE}
# First, assign a column in the pData table to umap pseudotime
pData(input_cds_lin)$Pseudotime <- pseudotime(input_cds_lin)

pData(input_cds_lin)$cell_subtype <- cut(pseudotime(input_cds_lin), 10)
binned_input_lin <- aggregate_by_cell_bin(input_cds_lin, "cell_subtype")
```

We are now ready to run Monocle 3's fit_models function to find sites 
that are differentially accessible across pseudotime. In this example, we 
include num_genes_expressed as a covariate to subtract its effect.

```{r, eval=TRUE}

# run 1000 random sites for speed
set.seed(1000)
acc_fits <- fit_models(binned_input_lin[sample(1:nrow(fData(binned_input_lin)), 1000),], model_formula_str = "~Pseudotime + num_genes_expressed" )

fit_coefs <- coefficient_table(acc_fits)

# Subset out the differentially accessible sites with respect to Pseudotime
pseudotime_terms <- subset(fit_coefs, term == "Pseudotime" & q_value < .05)

head(pseudotime_terms)
```

# References
Blondel, V.D., Guillaume, J.-L., Lambiotte, R., and Lefebvre, E. (2008). Fast unfolding of communities in large networks.

Dekker, J., Marti-Renom, M.A., and Mirny, L.A. (2013). Exploring the three-dimensional organization of genomes: interpreting chromatin interaction data. Nat. Rev. Genet. 14, 390–403.

Sanborn, A.L., Rao, S.S.P., Huang, S.-C., Durand, N.C., Huntley, M.H., Jewett, A.I., Bochkov, I.D., Chinnappan, D., Cutkosky, A., Li, J., et al. (2015). Chromatin extrusion explains key features of loop and domain formation in wild-type and engineered genomes. . Proc. Natl. Acad. Sci. U. S. A. 112, E6456–E6465.

Sexton, T., Yaffe, E., Kenigsberg, E., Bantignies, F., Leblanc, B., Hoichman, M., Parrinello, H., Tanay, A., and Cavalli, G. (2012). Three-Dimensional Folding and Functional Organization Principles of the Drosophila Genome. . Cell 148:3, 458-472.


# Citation
```{r}
citation("cicero")
```

# Session Info
```{r}
sessionInfo()
```

